### PR TITLE
Simplify Firestore Change Observers

### PR DESCRIPTION
**Pull request by rorybainfreetrade**
Currently we make use the same change observers for both Firestore and RTDB. However, Firestore only supports observing documents, not subpaths within documents. This means we unnecessarily traverse every key within each document written to Firestore, when instead we can just do simple checks like 'did this document exist before'.

This PR adds Firestore specific change observers that do these much simpler checks and should avoid extraneous work traversing each document.

### CODE REVIEW

**Comment by LukeInkster** [comment]
Was the title supposed to say "simplify"?

**Comment by rorybainfreetrade** [comment]
> Was the title supposed to say "simplify"?

Yes, yes it was 🤦

File: tests/driver/Firestore/Trigger/InProcessFirestore.onDelete.test.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            .onDelete(async (snapshot, context) => {
+                receivedSnapshots.push(snapshot)
+            })
+
+        // When a doc is created in that collection;
+        await driver
+            .firestore()
+            .collection("animals")
+            .doc("tiger")
+            .set({ colour: "orange", size: "large" })
```
Comment (LukeInkster) [inline tests/driver/Firestore/Trigger/InProcessFirestore.onDelete.test.ts:53]: Have you run a formatter over this? On line 27 we have exactly the same thing but it's split over 3 lines

File: tests/driver/Firestore/Trigger/InProcessFirestore.onUpdate.test.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        // And Firebase finishes its jobs;
+        await driver.jobsComplete()
+
+        // Then the handler should be triggered.
+        expect(receivedSnapshots).toHaveLength(1)
+        expect(receivedSnapshots[0].after!.data()).toEqual({ colour: "red", size: "large" })
+    })
+
+
+})
```
Comment (LukeInkster) [inline tests/driver/Firestore/Trigger/InProcessFirestore.onUpdate.test.ts:101]: Looks like this file needs a format too

File: src/driver/ChangeObserver/DatabaseChangeFilter.ts [hunk 3]
```diff
@@ -281,6 +354,39 @@ export class DeletedChangeFilter extends ChangeFilter {
...
+            return []
+        }
+
+        return [
+            {
+                parameters: matchPath.parameters,
+                change: {
+                    before: change.before,
+                    after: undefined,
+                    data: change.after,
```
[the remark this answers is not in the document]
Comment (rorybainfreetrade) [reply src/driver/ChangeObserver/DatabaseChangeFilter.ts]: I've aligned this with the previous RTDB/Firestore deletion filter, which has undefined as the `after` and points to the `before` data for `data`. I'm not sure this is actually used downstream in the deletion case, but they're at least aligned now.

**Pull request merged**